packages/licenses/src/getPkgInfo.ts

```diff
@@ -0,0 +1,356 @@
+        // to read the contents of a license file
+      }
+    }
+  }
+
+  return { name: license ?? 'Unknown' }
+}
+
+/**
+ * Fetch a file by integrity id from the content-addressable store
+ * @param file the file to fetch
+ * @param opts the options
+ * @returns Promise<Buffer>
+ */
+async function readLicenseFileFromCafs (cafsDir: string, fileIntegrity: string) {
+  const fileName = getFilePathByModeInCafs(cafsDir, fileIntegrity, 0)
+
+  try {
+    const fileContents = await readFile(fileName)
+    return fileContents
+  } catch (err: any) { // eslint-disable-line
+    if (err.code === 'ENOENT') return undefined
+    throw err
+  }
+}
+
+/**
+ * Returns the index of files included in
+ * the package identified by the integrity id
+ * @param packageIntegrity the integrity id of package
+ * @param opts options for fetching package file index
+ * @returns
+ */
+export async function readPackageIndexFile (
+  packageResolution: Resolution,
+  packageRef: string,
+  opts: { cafsDir: string, storeDir: string, lockfileDir: string }
+): Promise<
+  | {
+    local: false
+    files: Record<string, PackageFileInfo>
+  }
+  | {
+    local: true
+    files: Record<string, string>
+  }
+  > {
+  const isPackageWithIntegrity = 'integrity' in packageResolution
+
+  let pkgIndexFilePath
+  if (isPackageWithIntegrity) {
+    // Retrieve all the index file of all files included in the package
+    pkgIndexFilePath = await getFilePathInCafs(
+      opts.cafsDir,
+      packageResolution.integrity as string,
+      'index'
+    )
+  } else if (packageResolution.type === 'directory') {
+    // If the package resolution is of type 'directory' it means
+    // this is a workspace project and we need to do things differently
+    const targetRelative = depPathToFilename(packageRef)
+    const target = path.join(opts.storeDir, targetRelative)
+    pkgIndexFilePath = path.join(target, 'integrity.json')
+  } else if (!packageResolution.type && packageResolution.tarball) {
+    // If the package resolution has a tarball then we need to clean up
+    // the call to depPathToFilename as it adds '_[hash]' part to the
+    // directory for the package in the content-addressable store
+    const pathToDependency = depPathToFilename(packageRef)
+    const [packageDirInStore] = pathToDependency.split('_')
+    pkgIndexFilePath = path.join(
+      opts.storeDir,
+      packageDirInStore,
+      'integrity.json'
+    )
+  } else {
+    throw new PnpmError(
+      'UNSUPPORTED_PACKAGE_TYPE',
+      `Unsupported package resolution type for ${packageRef}`
+    )
+  }
+
+  // If the package resolution is of type directory we need to do things
+  // differently and generate our own package index file
+  const isLocalPkg = packageResolution.type === 'directory'
+  if (isLocalPkg) {
+    const localInfo = await fetchFromDir(
+      path.join(opts.lockfileDir, packageResolution.directory),
+      {}
+    )
+    return {
+      local: true,
+      files: localInfo.filesIndex,
+    }
```

Comment:
This will scan the whole package with all subdirectories even though the license files are only searched for in the root directory of the package.

Reply:
I couldn't find a way to get the package file index of a pnpm workspace project, so generating one. Is there a way to get it? It's not available in the CAFS